identity_stronghold/src/stronghold_jwk_storage_ext.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020-2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Wrapper around [`StrongholdSecretManager`](StrongholdSecretManager).
+
+use async_trait::async_trait;
+use identity_storage::key_storage::bls::*;
+use identity_storage::key_storage::JwkStorage;
+use identity_storage::JwkGenOutput;
+use identity_storage::JwkStorageBbsPlusExt;
+use identity_storage::KeyId;
+use identity_storage::KeyStorageError;
+use identity_storage::KeyStorageErrorKind;
+use identity_storage::KeyStorageResult;
+use identity_storage::KeyType;
+use identity_storage::ProofUpdateCtx;
+use identity_verification::jwk::Jwk;
+use iota_stronghold::procedures::FatalProcedureError;
+use iota_stronghold::procedures::Products;
+use iota_stronghold::procedures::Runner as _;
+use iota_stronghold::Location;
+use jsonprooftoken::jpa::algs::ProofAlgorithm;
+use std::str::FromStr;
+use zeroize::Zeroizing;
+use zkryptium::bbsplus::keys::BBSplusSecretKey;
+use zkryptium::bbsplus::signature::BBSplusSignature;
+
+use crate::stronghold_key_type::*;
+use crate::utils::*;
+use crate::StrongholdStorage;
+
+#[cfg_attr(not(feature = "send-sync-storage"), async_trait(?Send))]
+#[cfg_attr(feature = "send-sync-storage", async_trait)]
+impl JwkStorageBbsPlusExt for StrongholdStorage {
+  async fn generate_bbs(&self, key_type: KeyType, alg: ProofAlgorithm) -> KeyStorageResult<JwkGenOutput> {
+    let key_type = StrongholdKeyType::try_from(&key_type)?;
+
+    if !matches!(key_type, StrongholdKeyType::BLS12381G2) {
+      return Err(
+        KeyStorageError::new(KeyStorageErrorKind::UnsupportedKeyType)
+          .with_custom_message(format!("{key_type} is not supported")),
+      );
+    }
+
+    let stronghold = self.get_stronghold().await;
+    let client = get_client(&stronghold)?;
+
+    let kid: KeyId = random_key_id();
+    let target_key_location = Location::generic(
+      IDENTITY_VAULT_PATH.as_bytes().to_vec(),
+      kid.to_string().as_bytes().to_vec(),
+    );
+    let jwk = client
+      .exec_proc([], &target_key_location, |_| {
+        let (sk, pk) = generate_bbs_keypair(alg).map_err(|e| FatalProcedureError::from(e.to_string()))?;
+        let public_jwk = encode_bls_jwk(&sk, &pk, alg).1;
+
+        Ok(Products {
+          output: public_jwk,
+          secret: Zeroizing::new(sk.to_bytes().to_vec()),
+        })
+      })
+      .map_err(|e| {
+        KeyStorageError::new(KeyStorageErrorKind::Unspecified)
+          .with_custom_message("Failed to execute stronghold procedure")
+          .with_source(e)
+      })?;
+
+    persist_changes(self, stronghold).await?;
```

Comment:
I can't think of a solution besides putting the call to `persist_changes` inside `StrongholdStorage::drop` but that could require an async drop :'(. Another way could be to entrust this to the user but I don't really like the idea.

Reply:
I managed to work around the issue of async drop by spawning a `tokio::task` in there. Now all changes to `StrongholdStorage` are written to the snapshot file only when the storage handle is dropped instead of being written to at every operation.